vllm/model_executor/layers/quantization/gguf.py

```diff
@@ -0,0 +1,125 @@
+        if isinstance(layer, (LinearBase, VocabParallelEmbedding)):
+            return GGUFLinearMethod(self)
+        return None
+
+    def get_scaled_act_names(self) -> List[str]:
+        return []
+
+
+class GGUFLinearMethod(LinearMethodBase):
+    """Linear method for GGUF.
+
+    Args:
+        quant_config: The GGUF quantization config.
+    """
+
+    def __init__(self, quant_config: GGUFConfig):
+        self.quant_config = quant_config
+
+    def create_weights(self, layer: torch.nn.Module,
+                       input_size_per_partition: int,
+                       output_partition_sizes: List[int], input_size: int,
+                       output_size: int, params_dtype: torch.dtype,
+                       **extra_weight_attrs):
+
+        qweight = UninitializedParameter(requires_grad=False)
+        set_weight_attrs(qweight, {"input_dim": 1, "output_dim": 0})
+        set_weight_attrs(qweight, extra_weight_attrs)
+        layer.register_parameter("qweight", qweight)
+
+        qweight_type = Parameter(torch.empty(1, dtype=torch.uint8),
+                                 requires_grad=False)
+        set_weight_attrs(qweight_type, {
+            "is_gguf_weight_type": True,
+            "ignore_warning": True
+        })
+        set_weight_attrs(qweight_type, extra_weight_attrs)
+        layer.register_parameter("qweight_type", qweight_type)
+
+    def apply(self,
+              layer: torch.nn.Module,
+              x: torch.Tensor,
+              bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        qweight = layer.qweight
+        qweight_type = layer.qweight_type.item()
+        # use dequantize mulmat for IQmatrix, mmq for k-quants
+        if qweight_type >= 16:
+            block_size, type_size = GGML_QUANT_SIZES[qweight_type]
+            shape = (qweight.shape[0],
+                     qweight.shape[1] // type_size * block_size)
+            weight = ops.ggml_dequantize(qweight.contiguous(), qweight_type,
+                                         *shape)
+            out = x @ weight.T
+        else:
+            out = ops.ggml_mul_mat_a8(qweight.contiguous(), x, qweight_type,
+                                      qweight.shape[0])
+        if bias is not None:
+            out.add_(bias)
+        return out
+
+    def apply_embeds(self, layer: torch.nn.Module,
+                     x: torch.Tensor) -> torch.Tensor:
+        qweight = layer.qweight
+        qweight_type = layer.qweight_type.item()
+
+        block_size, type_size = GGML_QUANT_SIZES[qweight_type]
+        hidden_size = qweight.shape[1] // type_size * block_size
+        if qweight_type < 2:
+            return torch.embedding(qweight, x)
+        x_flat = x.flatten()
+        quant = torch.index_select(qweight, dim=0, index=x_flat)
+        dequant = ops.ggml_dequantize(quant, qweight_type, hidden_size,
+                                      x_flat.shape[0])
+        return dequant.view(*x.shape, hidden_size)
```

Comment:
This is needed because the embeddings can be quantized? This should be moved to a specific method as noted above.

Reply:
Yes, the embeddings can be quantized in gguf. And seems that the current `VocalParallelEmbeddings ` can't handle quantized embeddings well.

My concern is that separate embedding into a different method simply may break the logits calculation for `tie_wor_embeddings` models, because they will call `linear_method.apply` from `VocabParallelEmbedding` to get logits. 

IMO, to apply quantized embeddings, we can add extra `embed_method` besides `linear_method`, so that the quantized `VocabParallelEmbedding` can handle both embedding forward and linear forward.